brainiak/eventseg/event.py

```diff
@@ -0,0 +1,356 @@
+from sklearn.utils.validation import check_is_fitted, check_array
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "EventSegment",
+]
+
+
+class EventSegment(BaseEstimator):
+    """Class for learning and representing an event segmentation of
+    continuous fMRI data
+
+    Parameters
+    ----------
+    n_events: int
+        Number of segments to learn
+
+    step_var: Callable[[int], float] : default 4 * (0.98 ** (step - 1))
+        The Gaussian variance to use during fitting, as a function of the
+        number of steps. Should decrease slowly over time.
+
+    n_iter: int : default 500
+        Maximum number of steps to run during fitting
+
+    Attributes
+    ----------
+    p_start, p_end: length n_events+1 ndarray
+        initial and final prior distributions over events
+
+    P: n_events+1 by n_events+1 ndarray
+        HMM transition matrix
+
+    ll_ : ndarray with length = number of training datasets
+        Log-likelihood for training datasets over the course of training
+
+    segments_:  list of (time by event) ndarrays
+        Learned (soft) segmentation for training datasets
+
+    event_var_ : float
+        Gaussian variance at the end of learning
+
+    event_pat_ : voxel by event ndarray
+        Learned mean patterns for each event
+    """
+
+    def _default_var_schedule(step):
+        return 4 * (0.98 ** (step - 1))
+
+    def __init__(self, n_events=2,
+                 step_var=_default_var_schedule,
+                 n_iter=500):
+        self.n_events = n_events
+        self.step_var = step_var
+        self.n_iter = n_iter
+
+    def fit(self, X, y=None):
+        """Learn a segmentation on training data
+
+        Fits event patterns and a segmentation to training data. After
+        running this function, the learned event patterns can be used to
+        segment other datasets using find_events
+
+        Parameters
+        ----------
+        X: time by voxel ndarray, or a list of such ndarrays
+            fMRI data to be segmented. If a list is given, then all datasets
+            are segmented simultaneously with the same event patterns
+
+        y: not used (added to comply with BaseEstimator definition)
+
+        Returns
+        -------
+        self: the EventSegment object
+        """
+
+        if type(X) is not list:
+            X = check_array(X)
+            X = [X]
+
+        n_train = len(X)
+        for i in range(n_train):
+            X[i] = X[i].T
+
+        n_dim = X[0].shape[0]
+        for i in range(n_train):
+            assert (X[i].shape[0] == n_dim)
+
+        # Double-check that data is z-scored in time
+        for i in range(n_train):
+            X[i] = stats.zscore(X[i], axis=1, ddof=1)
+
+        # Set up transition matrix, with final sink state
+        self.p_start = np.zeros((1, self.n_events + 1))
+        self.p_start[0, 0] = 1
+        self.P = np.vstack((np.hstack((
+            0.5 * np.diag(np.ones(self.n_events))
```

Comment:
There is no complaint in this comment. But I am curious whether the choice of transition probability in self.P influences the result.